packages/use-subscription/src/__tests__/useSubscription-test.internal.js

```diff
@@ -560,4 +563,68 @@ describe('useSubscription', () => {
     act(() => renderer.update(<Subscription subscription={subscription2} />));
     Scheduler.unstable_flushAll();
   });
+
+  it('should not tear if a mutation occurs during a concurrent update', () => {
+    const input = document.createElement('input');
+
+    const mutate = value => {
+      input.value = value;
+      input.dispatchEvent(new Event('change'));
+    };
+
+    const subscription = {
+      getCurrentValue: () => input.value,
+      subscribe: callback => {
+        input.addEventListener('change', callback);
+        return () => input.removeEventListener('change', callback);
+      },
+    };
+
+    const Subscriber = ({id}) => {
+      const value = useSubscription(subscription);
+      Scheduler.unstable_yieldValue(`render:${id}:${value}`);
+      return value;
+    };
+
+    act(() => {
+      // Initial render of "A"
+      mutate('A');
+      ReactTestRenderer.create(
+        <React.Fragment>
+          <Subscriber id="first" />
+          <Subscriber id="second" />
+        </React.Fragment>,
+        {unstable_isConcurrent: true},
+      );
+      expect(Scheduler).toFlushAndYield(['render:first:A', 'render:second:A']);
+
+      // Partial update "A" -> "B"
+      // Interrupt with a second mutation "B" -> "C" but this should not cause tearing.
+      mutate('B');
+      expect(Scheduler).toFlushAndYieldThrough(['render:first:B']);
+      mutate('C');
+      expect(Scheduler).toFlushAndYield([
+        'render:second:B',
+        'render:first:C',
+        'render:second:C',
+      ]);
+
+      // No more pending updates
+      jest.runAllTimers();
+
+      // Partial update "C" -> "D"
+      // Interrupt with a second mutation "D" -> "E" but this should not cause tearing.
```

Comment:
This second interrupting mutation is required to trigger the previous failure. ([Described in more detail in this issue comment.](https://github.com/facebook/react/issues/16396#issuecomment-526684284)) I don't yet understand why this is but will be digging in more.

Reply:
It looks like React eagerly computes the state initially, for the first update. It doesn't do this or subsequent updates though because `expirationTime !== NoWork`.

I'm not sure if this is expected, given that my test flushes all pending work.

Reply:
@acdlite and I chatted about this offline and I have a better understanding of it now. I've updated the inline test comments. Going to hand this PR to him since he has full context now too.